custom_components/localtuya/fan.py

```diff
@@ -98,16 +106,26 @@ def supported_features(self) -> int:
 
     def status_updated(self):
         """Get state of Tuya fan."""
+        mappings = {
+            "1": [SPEED_LOW, 1],
+            "2": [SPEED_MEDIUM, 1],
+            "3": [SPEED_HIGH, 1],
+            "4": [SPEED_HIGH, 1],
+            "low": [SPEED_LOW, 2],
+            "medium": [SPEED_MEDIUM, 2],
+            "high": [SPEED_HIGH, 2],
+            "auto": [SPEED_LOW, 2],
+        }
+
+        dps_id = "%s" % self._dps_id
+
         self._is_on = self._status["dps"]["1"]
-        if not self._status["dps"]["1"]:
-            self._speed = SPEED_OFF
-        elif self._status["dps"]["2"] == "1":
-            self._speed = SPEED_LOW
-        elif self._status["dps"]["2"] == "2":
-            self._speed = SPEED_MEDIUM
-        elif self._status["dps"]["2"] == "3":
-            self._speed = SPEED_HIGH
-        self._oscillating = self._status["dps"]["8"]
+
+        self._speed = mappings[self._status["dps"][dps_id]][0]
+        self._type = mappings[self._status["dps"][dps_id]][1]
+
+        if hasattr(self._status["dps"], "8"):
```

Comment:
This does not work. With current abstraction we can do this check properly (as `self._status["dps"]` might not be set here yet), we will have to do it properly when we have a config option. In that case we can check like this:

```python
if self.has_config(CONF_FAN_OSCILLATING):
    ...
```

This makes the assumption that if the user has picked a DP for the oscillating feature, it is supported by the device. I guess you will have to check something like:

```python
if "8" in self._status.get("dps", {}):
    ...
```
For now.